### PR TITLE
SIO Client Token Expiration Support

### LINKED ISSUE #18: Plugin crashes from time to time
Similar issue as described here:
https://github.com/intelsdi-x/snap-plugin-collector-scaleio/issues/3

After crash, I see in logs:
```
time="2017-12-12T09:25:59+02:00" level=error msg="collector run error" _module=scheduler-job block=run error="CollectMetrics call error : Error while parsing data from /api/types/StoragePool/instances: json: cannot unmarshal object into Go value of type []map[string]interface {}" job-type=collector 
time="2017-12-12T09:25:59+02:00" level=warning msg="Task failed" _block=spin _module=scheduler-task consecutive failure limit=-1 consecutive failures=1653 error="CollectMetrics call error : Error while parsing data from /api/types/StoragePool/instances: json: cannot unmarshal object into Go value of type []map[string]interface {}" task-id=06405472-dbc7-46a3-ae03-78b2b5f03e3e task-name=Task-06405472-dbc7-46a3-ae03-78b2b5f03e3e 
```

Task file:

```
{
    "version": 1,
    "schedule": {
        "type": "simple",
        "interval": "10s"
    },
    "max-failures": -1,
    "workflow": {
        "collect": {
            "metrics": {
                "/intel/scaleio/*/BackgroundScanCompareCount": {},
                "/intel/scaleio/*/BackgroundScannedInMB": {},
                "/intel/scaleio/*/activeBckRebuildCapacityInKb": {},
                "/intel/scaleio/*/activeFwdRebuildCapacityInKb": {},
                "/intel/scaleio/*/activeMovingCapacityInKb": {},
                "/intel/scaleio/*/activeMovingInBckRebuildJobs": {},
                "/intel/scaleio/*/activeMovingInFwdRebuildJobs": {},
                "/intel/scaleio/*/activeMovingInNormRebuildJobs": {},
                "/intel/scaleio/*/activeMovingInRebalanceJobs": {},
                "/intel/scaleio/*/activeMovingOutBckRebuildJobs": {},
                "/intel/scaleio/*/activeMovingOutFwdRebuildJobs": {},
                "/intel/scaleio/*/activeMovingOutNormRebuildJobs": {},
                "/intel/scaleio/*/activeMovingRebalanceJobs": {},
                "/intel/scaleio/*/activeNormRebuildCapacityInKb": {},
                "/intel/scaleio/*/activeRebalanceCapacityInKb": {},
                "/intel/scaleio/*/atRestCapacityInKb": {},
                "/intel/scaleio/*/bckRebuildCapacityInKb": {},
                "/intel/scaleio/*/bckRebuildReadBwc/numOccured": {},
                "/intel/scaleio/*/bckRebuildReadBwc/numSeconds": {},
                "/intel/scaleio/*/bckRebuildReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/bckRebuildWriteBwc/numOccured": {},
                "/intel/scaleio/*/bckRebuildWriteBwc/numSeconds": {},
                "/intel/scaleio/*/bckRebuildWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/capacityAvailableForVolumeAllocationInKb": {},
                "/intel/scaleio/*/capacityInUseInKb": {},
                "/intel/scaleio/*/capacityLimitInKb": {},
                "/intel/scaleio/*/degradedFailedCapacityInKb": {},
                "/intel/scaleio/*/degradedFailedVacInKb": {},
                "/intel/scaleio/*/degradedHealthyCapacityInKb": {},
                "/intel/scaleio/*/degradedHealthyVacInKb": {},
                "/intel/scaleio/*/failedCapacityInKb": {},
                "/intel/scaleio/*/failedVacInKb": {},
                "/intel/scaleio/*/fixedReadErrorCount": {},
                "/intel/scaleio/*/fwdRebuildCapacityInKb": {},
                "/intel/scaleio/*/fwdRebuildReadBwc/numOccured": {},
                "/intel/scaleio/*/fwdRebuildReadBwc/numSeconds": {},
                "/intel/scaleio/*/fwdRebuildReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/fwdRebuildWriteBwc/numOccured": {},
                "/intel/scaleio/*/fwdRebuildWriteBwc/numSeconds": {},
                "/intel/scaleio/*/fwdRebuildWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/inMaintenanceCapacityInKb": {},
                "/intel/scaleio/*/inMaintenanceVacInKb": {},
                "/intel/scaleio/*/inUseVacInKb": {},
                "/intel/scaleio/*/maxCapacityInKb": {},
                "/intel/scaleio/*/movingCapacityInKb": {},
                "/intel/scaleio/*/normRebuildCapacityInKb": {},
                "/intel/scaleio/*/normRebuildReadBwc/numOccured": {},
                "/intel/scaleio/*/normRebuildReadBwc/numSeconds": {},
                "/intel/scaleio/*/normRebuildReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/normRebuildWriteBwc/numOccured": {},
                "/intel/scaleio/*/normRebuildWriteBwc/numSeconds": {},
                "/intel/scaleio/*/normRebuildWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/numOfDevices": {},
                "/intel/scaleio/*/numOfMappedToAllVolumes": {},
                "/intel/scaleio/*/numOfSnapshots": {},
                "/intel/scaleio/*/numOfThickBaseVolumes": {},
                "/intel/scaleio/*/numOfThinBaseVolumes": {},
                "/intel/scaleio/*/numOfUnmappedVolumes": {},
                "/intel/scaleio/*/numOfVolumes": {},
                "/intel/scaleio/*/numOfVolumesInDeletion": {},
                "/intel/scaleio/*/numOfVtrees": {},
                "/intel/scaleio/*/pendingBckRebuildCapacityInKb": {},
                "/intel/scaleio/*/pendingFwdRebuildCapacityInKb": {},
                "/intel/scaleio/*/pendingMovingCapacityInKb": {},
                "/intel/scaleio/*/pendingMovingInBckRebuildJobs": {},
                "/intel/scaleio/*/pendingMovingInFwdRebuildJobs": {},
                "/intel/scaleio/*/pendingMovingInNormRebuildJobs": {},
                "/intel/scaleio/*/pendingMovingInRebalanceJobs": {},
                "/intel/scaleio/*/pendingMovingOutBckRebuildJobs": {},
                "/intel/scaleio/*/pendingMovingOutFwdRebuildJobs": {},
                "/intel/scaleio/*/pendingMovingOutNormrebuildJobs": {},
                "/intel/scaleio/*/pendingMovingRebalanceJobs": {},
                "/intel/scaleio/*/pendingNormRebuildCapacityInKb": {},
                "/intel/scaleio/*/pendingRebalanceCapacityInKb": {},
                "/intel/scaleio/*/primaryReadBwc/numOccured": {},
                "/intel/scaleio/*/primaryReadBwc/numSeconds": {},
                "/intel/scaleio/*/primaryReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/primaryReadFromDevBwc/numOccured": {},
                "/intel/scaleio/*/primaryReadFromDevBwc/numSeconds": {},
                "/intel/scaleio/*/primaryReadFromDevBwc/totalWeightInKb": {},
                "/intel/scaleio/*/primaryReadFromRmcacheBwc/numOccured": {},
                "/intel/scaleio/*/primaryReadFromRmcacheBwc/numSeconds": {},
                "/intel/scaleio/*/primaryReadFromRmcacheBwc/totalWeightInKb": {},
                "/intel/scaleio/*/primaryVacInKb": {},
                "/intel/scaleio/*/primaryWriteBwc/numOccured": {},
                "/intel/scaleio/*/primaryWriteBwc/numSeconds": {},
                "/intel/scaleio/*/primaryWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/protectedCapacityInKb": {},
                "/intel/scaleio/*/protectedVacInKb": {},
                "/intel/scaleio/*/rebalanceCapacityInKb": {},
                "/intel/scaleio/*/rebalanceReadBwc/numOccured": {},
                "/intel/scaleio/*/rebalanceReadBwc/numSeconds": {},
                "/intel/scaleio/*/rebalanceReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/rebalanceWriteBwc/numOccured": {},
                "/intel/scaleio/*/rebalanceWriteBwc/numSeconds": {},
                "/intel/scaleio/*/rebalanceWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/rfacheReadHit": {},
                "/intel/scaleio/*/rfcacheAvgReadTime": {},
                "/intel/scaleio/*/rfcacheAvgWriteTime": {},
                "/intel/scaleio/*/rfcacheIoErrors": {},
                "/intel/scaleio/*/rfcacheIosOutstanding": {},
                "/intel/scaleio/*/rfcacheIosSkipped": {},
                "/intel/scaleio/*/rfcacheReadMiss": {},
                "/intel/scaleio/*/rfcacheReadsFromCache": {},
                "/intel/scaleio/*/rfcacheReadsPending": {},
                "/intel/scaleio/*/rfcacheReadsReceived": {},
                "/intel/scaleio/*/rfcacheReadsSkipped": {},
                "/intel/scaleio/*/rfcacheReadsSkippedAlignedSizeTooLarge": {},
                "/intel/scaleio/*/rfcacheReadsSkippedHeavyLoad": {},
                "/intel/scaleio/*/rfcacheReadsSkippedInternalError": {},
                "/intel/scaleio/*/rfcacheReadsSkippedLockIos": {},
                "/intel/scaleio/*/rfcacheReadsSkippedLowResources": {},
                "/intel/scaleio/*/rfcacheReadsSkippedMaxIoSize": {},
                "/intel/scaleio/*/rfcacheReadsSkippedStuckIo": {},
                "/intel/scaleio/*/rfcacheSkippedUnlinedWrite": {},
                "/intel/scaleio/*/rfcacheSourceDeviceReads": {},
                "/intel/scaleio/*/rfcacheSourceDeviceWrites": {},
                "/intel/scaleio/*/rfcacheWriteMiss": {},
                "/intel/scaleio/*/rfcacheWritePending": {},
                "/intel/scaleio/*/rfcacheWritesReceived": {},
                "/intel/scaleio/*/rfcacheWritesSkippedCacheMiss": {},
                "/intel/scaleio/*/rfcacheWritesSkippedHeavyLoad": {},
                "/intel/scaleio/*/rfcacheWritesSkippedInternalError": {},
                "/intel/scaleio/*/rfcacheWritesSkippedLowResources": {},
                "/intel/scaleio/*/rfcacheWritesSkippedMaxIoSize": {},
                "/intel/scaleio/*/rfcacheWritesSkippedStuckIo": {},
                "/intel/scaleio/*/rmPendingAllocatedInKb": {},
                "/intel/scaleio/*/secondaryReadBwc/numOccured": {},
                "/intel/scaleio/*/secondaryReadBwc/numSeconds": {},
                "/intel/scaleio/*/secondaryReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/secondaryReadFromDevBwc/numOccured": {},
                "/intel/scaleio/*/secondaryReadFromDevBwc/numSeconds": {},
                "/intel/scaleio/*/secondaryReadFromDevBwc/totalWeightInKb": {},
                "/intel/scaleio/*/secondaryReadFromRmcacheBwc/numOccured": {},
                "/intel/scaleio/*/secondaryReadFromRmcacheBwc/numSeconds": {},
                "/intel/scaleio/*/secondaryReadFromRmcacheBwc/totalWeightInKb": {},
                "/intel/scaleio/*/secondaryVacInKb": {},
                "/intel/scaleio/*/secondaryWriteBwc/numOccured": {},
                "/intel/scaleio/*/secondaryWriteBwc/numSeconds": {},
                "/intel/scaleio/*/secondaryWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/semiProtectedCapacityInKb": {},
                "/intel/scaleio/*/semiProtectedVacInKb": {},
                "/intel/scaleio/*/snapCapacityInUseInKb": {},
                "/intel/scaleio/*/snapCapacityInUseOccupiedInKb": {},
                "/intel/scaleio/*/spareCapacityInKb": {},
                "/intel/scaleio/*/thickCapacityInUseInKb": {},
                "/intel/scaleio/*/thinCapacityAllocatedInKm": {},
                "/intel/scaleio/*/thinCapacityInUseInKb": {},
                "/intel/scaleio/*/totalReadBwc/numOccured": {},
                "/intel/scaleio/*/totalReadBwc/numSeconds": {},
                "/intel/scaleio/*/totalReadBwc/totalWeightInKb": {},
                "/intel/scaleio/*/totalWriteBwc/numOccured": {},
                "/intel/scaleio/*/totalWriteBwc/numSeconds": {},
                "/intel/scaleio/*/totalWriteBwc/totalWeightInKb": {},
                "/intel/scaleio/*/unreachableUnusedCapacityInKb": {},
                "/intel/scaleio/*/unusedCapacityInKb": {}
            },
            "config": {
                    "/intel/scaleio": {
                    "username": "<api_user>",
                    "password": "<and_pwd>",
                    "gateway": "https://<gw_ip>",
                    "verifySSL": false
                }
            },
            "publish": [
                        {
                            "plugin_name": "influxdb",
                            "config": {
                                "host": "127.0.0.1",
                                "port": 8086,
                                "database": "<some_db>",
                                "user": "admin",
                                "password": "<super_secret>",
                                "scheme": "http"
                    }
                }
            ]
        }
    }
}

```

I tried to "avoid" this crash with
`"max-failures": -1,`

No luck. 
Basically I'd like to collect cluster statistics with snap, send it to influxdb and visualize with grafana. Works well until plugin crash.

```
snapteld -v
snapteld version 2.0.0

cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core) 

kernel: 3.10.0-693.5.2.el7.x86_64
```

### PR DESCRIPTION
**Pull request by king-jam**
* Moved into own package for easier maintenance
* Encapsulated Authentication

Fixes https://github.com/intelsdi-x/snap-plugin-collector-scaleio/issues/18

Summary of changes:
- Made it so token expiration is tracked and we will obtain a new token as needed.
- Added some error handling in the rare case we do fail a request

How to verify it:
Reference: https://github.com/intelsdi-x/snap-plugin-collector-scaleio/pull/21

Testing done:
Reference: https://github.com/intelsdi-x/snap-plugin-collector-scaleio/pull/21
Also did testing with task setup for > 10 minute intervals which would cause token invalidation due to inactivity.

A picture of a snapping turtle (not required but encouraged):
TURTLE

### CODE REVIEW

File: scaleio/client/client.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	return nil
+}
+
+// Logout invalidates current token and cient session
+func (c *SIOClient) Logout() error {
+	loginURL := &url.URL{}
+	//Make a copy of the base URL
+	*loginURL = *c.address
+	loginURL.Path = "/api/logout"
+	req, _ := http.NewRequest("GET", loginURL.String(), nil)
```
Comment (sandlbn) [inline scaleio/client/client.go]: can you check for errors here ?

File: scaleio/client/client.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	return nil
+}
+
+// GetAPIResponse takes a path and returns the data into the provided object
+func (c *SIOClient) GetAPIResponse(path string, v interface{}) error {
+	c.updatelastAccessTime() // needed to coordinate client reauth
+	fullURL := &url.URL{}
+	*fullURL = *c.address
+	fullURL.Path = path
+	req, _ := http.NewRequest("GET", fullURL.String(), nil)
```
Comment (sandlbn) [inline scaleio/client/client.go]: and here.

**Review by sandlbn — CHANGES_REQUESTED** [review]
Looks good just small comment. Thanks !

**Comment by king-jam** [comment]
Updated all the locations where that wasn't being checked.

Additional Debug Testing Data:
Task Starting Place
```
ID 					 NAME 						 STATE 		 HIT 	 MISS 	 FAIL 	 CREATED 		 LAST FAILURE
fd32fa59-ef21-4363-bf32-34140a2d5342 	 Task-fd32fa59-ef21-4363-bf32-34140a2d5342 	 Running 	 105 	 0 	 0 	 7:54PM 1-17-2018
```
Task Status after 12 hours:
```
ID 					 NAME 						 STATE 		 HIT 	 MISS 	 FAIL 	 CREATED 		 LAST FAILURE
fd32fa59-ef21-4363-bf32-34140a2d5342 	 Task-fd32fa59-ef21-4363-bf32-34140a2d5342 	 Running 	 4K 	 0 	 0 	 7:54PM 1-17-2018
```
ScaleIO API Operation Logs showing logout/login action every 4hrs:
```
2018-01-17 23:54:23,661 [http-nio-443-exec-5] INFO  audit - 10.238.32.81:/api/logout
2018-01-17 23:54:23,683 [http-nio-443-exec-7] INFO  audit - 10.238.32.81:/api/login
--
2018-01-18 03:54:23,802 [http-nio-443-exec-1] INFO  audit - 10.238.32.81:/api/logout
2018-01-18 03:54:23,825 [http-nio-443-exec-9] INFO  audit - 10.238.32.81:/api/login
--
2018-01-18 07:54:23,524 [http-nio-443-exec-7] INFO  audit - 10.238.32.81:/api/logout
2018-01-18 07:54:23,554 [http-nio-443-exec-4] INFO  audit - 10.238.32.81:/api/login
```

**Comment by king-jam** [comment]
@sandlbn should be all set now.

There are 2 corner case conditions that I will generate a separate PR for later.
1) Leaky Clients, there is no task cancellation signal to force cached client cleanup. It is extremely rare to become an issue since folks don't typically have more than a couple ScaleIO systems but current the client cache mechanism needs an age out policy added.
2) There is a need to see if a client needs to be invalidated in the case of a task configuration update with new credentials/settings. The gateway key would collide and try to use the old credentials. A trivial psuedohashing algorithm could fix this.

**Review by sandlbn — APPROVED** [review]
LGTM. Thanks for support ! Really appreciate. For now, we can document these 2 corner cases. I will generate new binaries after that.

**Pull request merged**